File: apps/api/app/schemas/feedback.py
"""Pydantic schemas for the /api/v1/feedback endpoints."""

from __future__ import annotations

from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel

FeedbackLabel = Literal[
    "correct_spam",
    "correct_safe",
    "false_positive",
    "false_negative",
    "not_sure",
]

RuleType = Literal["trust_sender", "block_sender", "trust_domain", "block_domain"]


class RuleSuggestion(BaseModel):
    type: RuleType
    value: str  # sender address or domain
    suggested: str  # human-readable reason


class FeedbackRequest(BaseModel):
    history_id: str
    feedback_label: FeedbackLabel
    reason: Optional[str] = None


class FeedbackResponse(BaseModel):
    success: bool
    feedback_id: UUID
    rule_suggestion: Optional[RuleSuggestion] = None

File: apps/web/lib/api/feedback.ts
const API_BASE = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

export type FeedbackLabel =
  | "correct_spam"
  | "correct_safe"
  | "false_positive"
  | "false_negative"
  | "not_sure";

export interface RuleSuggestion {
  type: string;
  sender?: string;
  domain?: string;
  suggested: boolean;
}

export interface SubmitFeedbackResponse {
  success: boolean;
  feedback_id: string;
  rule_suggestion: RuleSuggestion | null;
}

export async function submitFeedback(
  historyId: string,
  label: FeedbackLabel,
  reason?: string
): Promise<SubmitFeedbackResponse> {
  const res = await fetch(`${API_BASE}/api/v1/feedback`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    credentials: "include",
    body: JSON.stringify({
      history_id: historyId,
      feedback_label: label,
      reason,
    }),
  });
  if (!res.ok) throw new Error(`Failed to submit feedback (${res.status})`);
  return res.json();
}

export async function deleteFeedback(feedbackId: string): Promise<void> {
  const res = await fetch(`${API_BASE}/api/v1/feedback/${feedbackId}`, {
    method: "DELETE",
    credentials: "include",
  });
  if (!res.ok) throw new Error(`Failed to delete feedback (${res.status})`);
}
